File: mac/Sources/SuperconnectCore/FrameCodec.swift
import Foundation

/// Logical multiplexing channels. See proto/protocol.md §2.
public enum Channel: UInt8 {
    case control = 0
    case video = 1
    case input = 2
    case audio = 3
    case stats = 4
}

/// Per-frame flag bits. See proto/protocol.md §3.
public struct FrameFlags: OptionSet, Sendable {
    public let rawValue: UInt8
    public init(rawValue: UInt8) { self.rawValue = rawValue }
    public static let keyframe = FrameFlags(rawValue: 0x01)
    public static let codecConfig = FrameFlags(rawValue: 0x02)
}

/// One protocol message: a channel, flags, and a payload.
public struct Frame: Equatable {
    public let channel: UInt8
    public let flags: UInt8
    public let payload: Data

    public init(channel: UInt8, flags: UInt8 = 0, payload: Data) {
        self.channel = channel
        self.flags = flags
        self.payload = payload
    }

    public init(channel: Channel, flags: FrameFlags = [], payload: Data) {
        self.init(channel: channel.rawValue, flags: flags.rawValue, payload: payload)
    }
}

/// Errors raised while decoding a byte stream into frames.
public enum FrameError: Error, Equatable {
    case payloadTooLarge(Int)
}

/// Wire framing: `channel:u8 | flags:u8 | length:u32-LE | payload`.
/// This is the single most important interop surface — it must match
/// proto/vectors.json byte-for-byte across Swift, C++, and ArkTS.
public enum FrameCodec {
    public static let headerSize = 6
    public static let maxPayload = 16 * 1024 * 1024 // 16 MiB

    public static func encode(channel: UInt8, flags: UInt8, payload: Data) -> Data {
        var out = Data(capacity: headerSize + payload.count)
        out.append(channel)
        out.append(flags)
        let len = UInt32(payload.count)
        out.append(UInt8(len & 0xff))
        out.append(UInt8((len >> 8) & 0xff))
        out.append(UInt8((len >> 16) & 0xff))
        out.append(UInt8((len >> 24) & 0xff))
        out.append(payload)
        return out
    }

    public static func encode(_ frame: Frame) -> Data {
        encode(channel: frame.channel, flags: frame.flags, payload: frame.payload)
    }

    public static func encode(channel: Channel, flags: FrameFlags = [], payload: Data) -> Data {
        encode(channel: channel.rawValue, flags: flags.rawValue, payload: payload)
    }
}

/// Accumulates incoming bytes and yields complete frames as they arrive.
/// Tolerant of arbitrary fragmentation (one frame split across many reads,
/// or many frames in one read).
public final class FrameDecoder {
    private var buffer = [UInt8]()

    public init() {}

    public func push(_ data: Data) throws -> [Frame] {
        buffer.append(contentsOf: data)
        var frames = [Frame]()
        var cursor = 0
        while buffer.count - cursor >= FrameCodec.headerSize {
            let base = cursor
            let channel = buffer[base]
            let flags = buffer[base + 1]
            let len = Int(buffer[base + 2])
                | (Int(buffer[base + 3]) << 8)
                | (Int(buffer[base + 4]) << 16)
                | (Int(buffer[base + 5]) << 24)
            if len > FrameCodec.maxPayload {
                throw FrameError.payloadTooLarge(len)
            }
            let total = FrameCodec.headerSize + len
            if buffer.count - base < total { break } // wait for more bytes
            let payloadStart = base + FrameCodec.headerSize
            let payload = Data(buffer[payloadStart ..< payloadStart + len])
            frames.append(Frame(channel: channel, flags: flags, payload: payload))
            cursor += total
        }
        if cursor > 0 { buffer.removeFirst(cursor) }
        return frames
    }
}

// MARK: - Hex helpers (used by tests and logging)

public extension Data {
    init(hexString: String) {
        var data = Data()
        var idx = hexString.startIndex
        while idx < hexString.endIndex {
            let next = hexString.index(idx, offsetBy: 2, limitedBy: hexString.endIndex) ?? hexString.endIndex
            if let byte = UInt8(hexString[idx ..< next], radix: 16) { data.append(byte) }
            idx = next
        }
        self = data
    }

    var hexEncodedString: String {
        map { String(format: "%02x", $0) }.joined()
    }
}

File: android/protocol/src/main/kotlin/com/superconnect/protocol/FrameCodec.kt
package com.superconnect.protocol

/**
 * Wire framing — Android/Kotlin mirror of mac SuperconnectCore/FrameCodec.swift (and the ArkTS / C++
 * impls). Pure Kotlin (no Android deps) so it lives in a JVM library module and is conformance-tested
 * against proto/vectors.json off-device. Layout: `channel:u8 | flags:u8 | length:u32-LE | payload`.
 * This is the single most important interop surface — it must stay byte-identical across all platforms.
 */

/** Logical multiplexing channels. proto/protocol.md §2. */
enum class Channel(val value: Int) {
    CONTROL(0), VIDEO(1), INPUT(2), AUDIO(3), STATS(4);
    companion object { fun from(v: Int): Channel? = entries.firstOrNull { it.value == v } }
}

/** Per-frame flag bits. proto/protocol.md §3. */
object FrameFlags {
    const val KEYFRAME = 0x01
    const val CODEC_CONFIG = 0x02
}

/** One protocol message: a channel, flags, and a payload. */
class Frame(val channel: Int, val flags: Int, val payload: ByteArray) {
    override fun equals(other: Any?): Boolean =
        other is Frame && channel == other.channel && flags == other.flags && payload.contentEquals(other.payload)
    override fun hashCode(): Int = (channel * 31 + flags) * 31 + payload.contentHashCode()
}

object FrameCodec {
    const val HEADER_SIZE = 6
    const val MAX_PAYLOAD = 16 * 1024 * 1024   // 16 MiB

    fun encode(channel: Int, flags: Int, payload: ByteArray): ByteArray {
        val out = ByteArray(HEADER_SIZE + payload.size)
        out[0] = channel.toByte()
        out[1] = flags.toByte()
        val len = payload.size
        out[2] = (len and 0xff).toByte()
        out[3] = ((len ushr 8) and 0xff).toByte()
        out[4] = ((len ushr 16) and 0xff).toByte()
        out[5] = ((len ushr 24) and 0xff).toByte()
        payload.copyInto(out, HEADER_SIZE)
        return out
    }

    fun encode(channel: Channel, flags: Int, payload: ByteArray): ByteArray =
        encode(channel.value, flags, payload)
}

/**
 * Accumulates incoming bytes and yields complete frames as they arrive — tolerant of arbitrary
 * fragmentation (one frame split across reads, or many frames in one read). Mirrors the Swift decoder.
 */
class FrameDecoder {
    private var buffer = ByteArray(0)

    fun push(data: ByteArray): List<Frame> {
        buffer += data
        val frames = ArrayList<Frame>()
        var cursor = 0
        while (buffer.size - cursor >= FrameCodec.HEADER_SIZE) {
            val b = cursor
            val channel = buffer[b].toInt() and 0xff
            val flags = buffer[b + 1].toInt() and 0xff
            // Decode the u32 length as UNSIGNED (accumulate in a Long). A Kotlin Int is signed, so a length
            // with bit 31 set (e.g. 0xFFFFFFFF) would read negative and skip the MAX_PAYLOAD guard below —
            // diverging from the Swift/C++/ArkTS codecs, which all treat it unsigned and reject >16 MiB as a
            // protocol error (proto/protocol.md). Keeping it unsigned makes the four codecs byte-identical.
            val len = (buffer[b + 2].toLong() and 0xff) or
                ((buffer[b + 3].toLong() and 0xff) shl 8) or
                ((buffer[b + 4].toLong() and 0xff) shl 16) or
                ((buffer[b + 5].toLong() and 0xff) shl 24)
            if (len > FrameCodec.MAX_PAYLOAD) throw IllegalStateException("frame payload too large: $len")
            val payloadLen = len.toInt()   // safe: the guard above bounds len to <= 16 MiB
            val total = FrameCodec.HEADER_SIZE + payloadLen
            if (buffer.size - b < total) break   // wait for more bytes
            val start = b + FrameCodec.HEADER_SIZE
            frames.add(Frame(channel, flags, buffer.copyOfRange(start, start + payloadLen)))
            cursor += total
        }
        if (cursor > 0) buffer = buffer.copyOfRange(cursor, buffer.size)
        return frames
    }
}
